code/ship/ship.cpp

```diff
@@ -13159,15 +13160,59 @@ int get_available_secondary_weapons(object *objp, int *outlist, int *outbanklist
 	int	count = 0;
 	int	i;
 	ship	*shipp;
-
+	weapon_info *wepp;
 	Assert(objp->type == OBJ_SHIP);
 	Assert((objp->instance >= 0) && (objp->instance < MAX_SHIPS));
 	shipp = &Ships[objp->instance];
+	ai_info* aip = &Ai_info[shipp->ai_index];
+	
+	float target_range, weapon_range_max, weapon_range_min;
+	target_range = 0.0f;
 
+	if (The_mission.ai_profile->ai_range_secondary_select_mode!= AI_RANGE_SEC_SEL_MODE_RETAIL) {
+		vec3d our_position = objp->pos;
+		vec3d target_position;
+		object *target = &Objects[Ai_info[shipp->ai_index].target_objnum];
+		if (target->type == OBJ_SHIP) {
+			if (aip->targeted_subsys != NULL) {
+				get_subsystem_pos(&target_position, target, aip->targeted_subsys);
+			}
+			else if (Ship_info[shipp->ship_info_index].is_big_or_huge() ){
+				ai_big_pick_attack_point(target, objp, &target_position, 0.8f);
+			}
+			else {
+				target_position = target->pos;
+			}
+		}
+		else {
+			target_position = target->pos;
+		}
+		target_range = vm_vec_dist_quick(&our_position, &target_position);
+	}
 	for (i=0; i<shipp->weapons.num_secondary_banks; i++)
 		if (shipp->weapons.secondary_bank_ammo[i]) {
-			outbanklist[count] = i;
-			outlist[count++] = shipp->weapons.secondary_bank_weapons[i];
+			if (The_mission.ai_profile->ai_range_secondary_select_mode != AI_RANGE_SEC_SEL_MODE_RETAIL) {
+				wepp = &Weapon_info[shipp->weapons.secondary_bank_weapons[i]];
+				weapon_range_min = wepp->WeaponMinRange;
+				weapon_range_max = wepp->weapon_range;
+				//If weapon range is not set in the weapon info, derive it
+				if (weapon_range_max >= 999999999.0f) { 
```

Comment:
I think it would be better to pick a smaller number as the threshold.

Reply:
If memory serves, this number was chosen because it is the default if no range is set by the table. I don't want to ignore a modder's tabled value if they have one, even if it's an absurd one.

Reply:
Two reasons:

1) The original number was probably chosen arbitrarily.  There may be good reasons to override the range if it's very large but below the threshold.
2) The initialized number is "magic".  It isn't defined anywhere.  Suppose someone decides to change the initialized value but forgets to change it here too?

If there's a sensible threshold above which to derive the range, then do that.  Otherwise, change `999999999.0f` to a #define and use the #define in both places.

Reply:
I'll make the changeover of the magic number to a define, but unless someone can give me a concrete and compelling situation where it'd be beneficial to do so I can't see my way to overriding excessively long but user set ranges.

Reply:
I can't think of such a situation offhand, but I can't be sure that no situation exists.

Reply:
I'm pretty opposed to implementing any sort of override on principal, as it's my general stance that the engine should let the modder do things that seem absurd on the surface unless they are blatantly invalid and destabilizing. And wthout a specific case I can't very well pick a reasonable override to implement anyway, so I'm going to leave it as it is.

Reply:
Fair enough.